File: src/surface/ffi/cursor.zig
//! gist in-process FFI — the PULL-cursor surface.
//!
//! The rung-3 `open`/`search`/`close` triad (`session.zig`) streams matches to a
//! host callback (PUSH). This module is its pull sibling over the curated Zig
//! facade (`api.zig`): a host opens an `Engine`, runs a `search` that
//! materializes an owned `Cursor`, then walks it with `next` / `next_batch` —
//! inverting control so a caller that can't (or won't) yield its stack to a
//! callback (Go's cgo, an async runtime, a REPL) still drives the warm engine.
//!
//! Every entry returns a `contract.Status` and never `die()`s — the same
//! fail-closed property the search ABI is gated on. Handles are opaque and
//! owned: `Engine` and `CancelToken` are heap-boxed `api` values; a `Cursor`
//! wrapper owns the facade cursor plus a reusable submatch scratch. Each has one
//! NULL-safe destructor.
//!
//! ## View lifetime
//!
//! A `contract.Match` filled by `next` / `next_batch` BORROWS: its `path` and
//! `line` alias the cursor's arena (valid until `cursorClose`), and its
//! `submatches` alias the wrapper's reusable scratch (valid only until the next
//! `next` / `next_batch` on that cursor). The host copies anything it keeps —
//! exactly the discipline the callback path already documents.

const std = @import("std");
const api = @import("irregex").api;
const contract = @import("contract.zig");

/// Share the host's C heap so a boxed handle frees without a Zig GPA.
const gpa = std.heap.c_allocator;
const Status = contract.Status;

/// A pull cursor: the facade cursor (owns the records) plus a reusable submatch
/// buffer the view structs point into. One `subs` scratch is refilled per
/// `next`; `nextBatch` fills it once for the whole batch (offsets computed so a
/// realloc can't dangle a mid-batch view).
pub const Cursor = struct {
    inner: *api.Cursor,
    subs: std.ArrayList(contract.Submatch) = .empty,

    fn deinit(self: *Cursor) void {
        self.subs.deinit(gpa);
        self.inner.deinit();
    }
};

// The engine a cursor is drawn from, and the token that cancels it, are opened
// by `libirregex` (`irregex_engine_open` / `irregex_cancel_new`) rather than
// here. Every package's producer takes one, and an engine can only be read by
// the copy of the engine code that made it — so the opener is substrate, and
// this file owns only what search does WITH the corpus it is handed.

/// Run one search and materialize a pull cursor. Null / wrongly-sized / unknown
/// options fail closed with `.invalid`; an unsupported pattern returns `.stale`
/// (answer cold), never a cursor.
pub fn searchCursor(engine: *api.Engine, req_ptr: ?*const contract.SearchRequest, out: ?**Cursor) Status {
    contract.beginCall();
    const slot = out orelse return .invalid;
    const req = req_ptr orelse return .invalid;
    if (req.struct_size != @sizeOf(contract.SearchRequest) or req.flags & ~contract.known_flags != 0)
        return .invalid;
    const pattern: []const u8 = if (req.pattern_len == 0) "" else blk: {
        const p = req.pattern orelse return .invalid;
        break :blk p[0..req.pattern_len];
    };
    const flags = req.flags;
    const query = api.SearchQuery{
        .pattern = pattern,
        .fixed = flags & contract.flag_fixed != 0,
        .ignore_case = flags & contract.flag_ignore_case != 0,
        .smart_case = flags & contract.flag_smart_case != 0,
        .word = flags & contract.flag_word != 0,
        .invert = flags & contract.flag_invert != 0,
        .unicode = flags & contract.flag_no_unicode == 0,
        .before = req.before_context,
        .after = req.after_context,
        .max_count = if (flags & contract.flag_max_count != 0) req.max_count else null,
    };
    const run = api.RunOptions{
        .cancel = req.cancel,
        .timeout_ns = if (req.timeout_ns == 0) null else req.timeout_ns,
        .max_results = if (req.max_results == 0) null else req.max_results,
    };
    const answered = engine.search(query, run) catch |e| switch (e) {
        error.OutOfMemory => return contract.report(.{ .code = error.OutOfMemory }),
    };
    const inner = switch (answered) {
        .got => |c| c,
        // A declinature — the caller answers cold and gets the identical
        // result, so nothing lands in the fault slot.
        .declined => return .stale,
    };
    const handle = gpa.create(Cursor) catch {
        inner.deinit();
        return contract.report(.{ .code = error.OutOfMemory });
    };
    handle.* = .{ .inner = inner };
    slot.* = handle;
    return .ok;
}

/// Fill `out` with the next record. Returns `.match` (a record was written),
/// `.ok` (end of stream — `out` untouched), or `.invalid` / `.out_of_memory`.
pub fn cursorNext(cursor: *Cursor, out: ?*contract.Match) Status {
    contract.beginCall();
    const slot = out orelse return .invalid;
    const rec = cursor.inner.next() orelse return .ok;
    cursor.subs.clearRetainingCapacity();
    cursor.subs.ensureTotalCapacity(gpa, rec.spans.len) catch return contract.report(.{ .code = error.OutOfMemory });
    for (rec.spans) |sp| cursor.subs.appendAssumeCapacity(spanView(rec.text, sp));
    slot.* = viewOf(rec, cursor.subs.items.ptr, cursor.subs.items.len);
    return .match;
}

/// Fill up to `cap` records into `out[0..cap]`; writes the count to `written`.
/// Returns `.match` when ≥1 record was written, `.ok` at end of stream (0
/// written), or a failure. All views in the batch share the wrapper's scratch,
/// so the batch is valid only until the next call on this cursor.
pub fn cursorNextBatch(cursor: *Cursor, out_ptr: ?[*]contract.Match, cap: usize, written: ?*usize) Status {
    contract.beginCall();
    const count_slot = written orelse return .invalid;
    count_slot.* = 0;
    if (cap == 0) return .ok;
    const out = (out_ptr orelse return .invalid)[0..cap];

    // Gather the batch of owned records first (the facade cursor advances), so
    // one pass sizes the shared submatch scratch before any view points into it
    // — a mid-batch realloc then can't dangle an already-written view.
    var gathered = std.ArrayList(api.OwnedMatch).empty;
    defer gathered.deinit(gpa);
    var total_spans: usize = 0;
    while (gathered.items.len < cap) {
        const rec = cursor.inner.next() orelse break;
        gathered.append(gpa, rec) catch return contract.report(.{ .code = error.OutOfMemory });
        total_spans += rec.spans.len;
    }
    if (gathered.items.len == 0) return .ok;

    cursor.subs.clearRetainingCapacity();
    cursor.subs.ensureTotalCapacity(gpa, total_spans) catch return contract.report(.{ .code = error.OutOfMemory });
    var offset: usize = 0;
    for (gathered.items, 0..) |rec, i| {
        for (rec.spans) |sp| cursor.subs.appendAssumeCapacity(spanView(rec.text, sp));
        out[i] = viewOf(rec, cursor.subs.items.ptr + offset, rec.spans.len);
        offset += rec.spans.len;
    }
    count_slot.* = gathered.items.len;
    return .match;
}

/// Whether any file matched (cold's exit-code boolean), even if a budget cut the
/// scan short. 1 = matched, 0 = none.
pub fn cursorMatched(cursor: *Cursor) i32 {
    return @intFromBool(cursor.inner.anyMatched());
}

/// Free a cursor (and its record buffer + scratch).
pub fn cursorClose(cursor: *Cursor) void {
    cursor.deinit();
    gpa.destroy(cursor);
}

fn spanView(text: []const u8, sp: api.Span) contract.Submatch {
    return .{ .text = text.ptr + sp.start, .len = sp.end - sp.start, .start = sp.start, .end = sp.end };
}

fn viewOf(rec: api.OwnedMatch, subs: [*]const contract.Submatch, nsubs: usize) contract.Match {
    return .{
        .path = rec.path.ptr,
        .path_len = rec.path.len,
        .line_number = rec.line_number,
        .line = rec.text.ptr,
        .line_len = rec.text.len,
        .submatches = subs,
        .nsubmatches = nsubs,
        .kind = @enumFromInt(@intFromEnum(rec.kind)),
    };
}

File: bindings/rust/src/exact/cursor.rs
//! The in-process `Engine` / `Cursor` surface over the pull-cursor C ABI
//! (ADR-352), compiled only under the `native` feature.
//!
//! The crate's top-level [`crate::search`] helpers answer a *one-shot* query over
//! the certified subprocess. This module is the other shape a host wants: a **warm
//! [`Engine`] held open** across many queries, each producing a **pull [`Cursor`]**
//! the caller drives at its own pace — the callback-free sibling of the push
//! session, so no C-to-Rust trampoline runs during a scan.
//!
//! Ownership is RAII end to end. [`Engine`] owns the warm corpus + index + I/O
//! pool and serializes [`search`](Engine::search) (the resident engine is
//! single-writer), but a materialized [`Cursor`] owns its records in a private
//! arena and is fully independent of the engine — cursors outlive it and iterate
//! without the engine lock. [`CancelToken`] is an explicit, thread-safe stop one
//! thread trips while another is blocked in `search`. Every handle frees on
//! `Drop`; failures are the crate's typed [`Error`], never a terminated host.

use std::ffi::CString;
use std::mem::MaybeUninit;
use std::path::Path;
use std::sync::Mutex;
use std::time::Duration;

use crate::contract::{Match, MatchKind, Submatch};
use crate::error::{Error, Result};
use crate::request::{SearchEngine, SearchRequest};
use crate::sys;

/// Records-per-native-call default for [`Cursor::batches`]: enough to amortize the
/// FFI crossing without holding a large transient view buffer.
pub const DEFAULT_BATCH: usize = 64;

/// A thread-safe cooperative stop shared into [`Engine`] searches via [`Run::cancel`].
///
/// One thread may [`cancel`](Self::cancel) while another is blocked in
/// [`Engine::search`] (the native scan holds no Rust lock the canceler needs);
/// the scan stops at its next record boundary, keeping whatever it gathered. A
/// token is reusable across searches until dropped.
pub struct CancelToken {
    inner: *mut sys::irregex_cancel,
}

// The token is a bare atomic flag behind the pointer; sharing `&CancelToken`
// across threads (to cancel a scan another thread runs) is the whole point.
unsafe impl Send for CancelToken {}
unsafe impl Sync for CancelToken {}

impl CancelToken {
    /// Allocate a fresh (unset) token.
    ///
    /// # Errors
    /// [`Error::Failed`] if the native allocation fails (out of memory).
    pub fn new() -> Result<Self> {
        let mut out: *mut sys::irregex_cancel = std::ptr::null_mut();
        let status = unsafe { sys::irregex_cancel_new(&mut out) };
        if status != sys::OK {
            return Err(status_error(status, "allocate cancel token"));
        }
        Ok(Self { inner: out })
    }

    /// Request cancellation of any in-flight search using this token.
    pub fn cancel(&self) {
        unsafe { sys::irregex_cancel_request(self.inner) };
    }
}

impl Drop for CancelToken {
    fn drop(&mut self) {
        unsafe { sys::irregex_cancel_free(self.inner) };
    }
}

/// Per-operation budgets for one [`Engine::search`], each honored at a record
/// boundary (never a torn record). All absent = run to completion.
#[derive(Default)]
pub struct Run<'a> {
    /// A stop another thread may trip mid-scan.
    pub cancel: Option<&'a CancelToken>,
    /// Wall-clock ceiling measured from the start of the search.
    pub timeout: Option<Duration>,
    /// Stop after this many records land.
    pub max_results: Option<usize>,
}

impl<'a> Run<'a> {
    /// Attach a cancellation token.
    #[must_use]
    pub fn cancel(mut self, token: &'a CancelToken) -> Self {
        self.cancel = Some(token);
        self
    }

    /// Set a wall-clock budget.
    #[must_use]
    pub fn timeout(mut self, d: Duration) -> Self {
        self.timeout = Some(d);
        self
    }

    /// Set a result-count budget.
    #[must_use]
    pub fn max_results(mut self, n: usize) -> Self {
        self.max_results = Some(n);
        self
    }
}

/// A pull result handle over one search: an iterator of owned [`Match`] records.
///
/// Iterate it directly ([`Iterator`]) for record-at-a-time consumption, or call
/// [`batches`](Self::batches) to amortize the native crossing. Records are copied
/// into owned values, so they outlive the cursor (and the engine); the native
/// record buffer frees on [`Drop`].
#[derive(Debug)]
pub struct Cursor {
    inner: *mut sys::irregex_cursor,
    matched: Option<bool>,
    done: bool,
}

// A cursor is a self-contained owned buffer; it may move between threads, but a
// single cursor is a single iterator (no `Sync`).
unsafe impl Send for Cursor {}

impl Cursor {
    /// Whether any file matched (cold's exit-code boolean), even if a budget cut
    /// the scan short. Stable for the cursor's lifetime.
    pub fn matched(&mut self) -> bool {
        *self
            .matched
            .get_or_insert_with(|| unsafe { sys::irregex_cursor_matched(self.inner) } != 0)
    }

    /// Yield lists of up to `size` records, each filled by one native call — the
    /// same records [`Iterator::next`] yields, chunked to trade per-record call
    /// overhead for a larger transient view buffer.
    ///
    /// # Panics
    /// If `size` is zero.
    pub fn batches(self, size: usize) -> Batches {
        assert!(size >= 1, "batch size must be >= 1");
        Batches {
            cursor: self,
            buf: (0..size).map(|_| MaybeUninit::uninit()).collect(),
        }
    }
}

impl Iterator for Cursor {
    type Item = Result<Match>;

    fn next(&mut self) -> Option<Self::Item> {
        if self.done {
            return None;
        }
        let mut view = MaybeUninit::<sys::MatchView>::uninit();
        let status = unsafe { sys::irregex_cursor_next(self.inner, view.as_mut_ptr()) };
        match status {
            sys::MATCH => Some(Ok(to_match(unsafe { view.assume_init_ref() }))),
            sys::OK => {
                self.done = true;
                None
            }
            other => {
                self.done = true;
                Some(Err(status_error(other, "cursor advance")))
            }
        }
    }
}

impl Drop for Cursor {
    fn drop(&mut self) {
        unsafe { sys::irregex_cursor_close(self.inner) };
    }
}

/// The chunked view of a [`Cursor`] returned by [`Cursor::batches`]: an iterator
/// of owned record lists, each filled by a single native `next_batch` call.
pub struct Batches {
    cursor: Cursor,
    buf: Vec<MaybeUninit<sys::MatchView>>,
}

impl Iterator for Batches {
    type Item = Result<Vec<Match>>;

    fn next(&mut self) -> Option<Self::Item> {
        if self.cursor.done {
            return None;
        }
        let mut written: usize = 0;
        let status = unsafe {
            sys::irregex_cursor_next_batch(
                self.cursor.inner,
                self.buf.as_mut_ptr().cast::<sys::MatchView>(),
                self.buf.len(),
                &mut written,
            )
        };
        match status {
            sys::MATCH => {
                let views = &self.buf[..written];
                Some(Ok(views
                    .iter()
                    .map(|v| to_match(unsafe { v.assume_init_ref() }))
                    .collect()))
            }
            sys::OK => {
                self.cursor.done = true;
                None
            }
            other => {
                self.cursor.done = true;
                Some(Err(status_error(other, "cursor batch")))
            }
        }
    }
}

/// A warm in-process corpus queried many times, each yielding a pull [`Cursor`].
///
/// Open it over zero or more roots ([`open`](Self::open); no roots = the rootless
/// CWD walk a bare `gist <pattern>` scans). [`search`](Self::search) is serialized
/// (the resident engine is single-writer), but the cursors it returns are
/// independent and iterable in parallel. The corpus frees on [`Drop`].
pub struct Engine {
    inner: *mut sys::irregex_engine,
    lock: Mutex<()>,
}

// The pointer is only touched under `lock`; a materialized cursor never reaches
// back into the engine, so `&Engine` is safe to share across threads.
unsafe impl Send for Engine {}
unsafe impl Sync for Engine {}

impl Engine {
    /// Open a warm engine over `roots` (each an absolute or CWD-relative path;
    /// none = the rootless CWD walk).
    ///
    /// # Errors
    /// [`Error::Failed`] if the corpus cannot be stood up (bad root, out of memory).
    pub fn open<I, P>(roots: I) -> Result<Self>
    where
        I: IntoIterator<Item = P>,
        P: AsRef<Path>,
    {
        let owned: Vec<CString> = roots
            .into_iter()
            .map(|p| cstring(p.as_ref()))
            .collect::<Result<_>>()?;
        let ptrs: Vec<*const std::os::raw::c_char> = owned.iter().map(|c| c.as_ptr()).collect();
        let root_ptr = if ptrs.is_empty() {
            std::ptr::null()
        } else {
            ptrs.as_ptr()
        };
        let mut out: *mut sys::irregex_engine = std::ptr::null_mut();
        let status = unsafe { sys::irregex_engine_open(root_ptr, ptrs.len(), &mut out) };
        if status != sys::OK {
            return Err(status_error(status, "engine open"));
        }
        Ok(Self {
            inner: out,
            lock: Mutex::new(()),
        })
    }

    /// A fresh cancellation token for use with [`Run::cancel`].
    ///
    /// # Errors
    /// [`Error::Failed`] if the native allocation fails.
    pub fn cancel_token(&self) -> Result<CancelToken> {
        CancelToken::new()
    }

    /// Search over the warm corpus, returning a pull [`Cursor`], with default
    /// (unbounded) budgets. `request.paths` and its subprocess-only `timeout` are
    /// ignored — roots belong to the engine ([`open`](Self::open)).
    ///
    /// # Errors
    /// [`Error::UnsupportedPattern`] for a pattern outside the linear engine,
    /// [`Error::Unrepresentable`] for an option the in-process ABI cannot honor,
    /// [`Error::Failed`] on a native failure.
    pub fn search(&self, request: &SearchRequest) -> Result<Cursor> {
        self.run(request, Run::default())
    }

    /// Search with explicit per-operation budgets ([`Run`]).
    ///
    /// # Errors
    /// See [`search`](Self::search).
    pub fn run(&self, request: &SearchRequest, run: Run<'_>) -> Result<Cursor> {
        reject_unrepresentable(request)?;
        let pattern = request.pattern.as_bytes();
        let (before, after) = if request.before != 0 || request.after != 0 {
            (u64::from(request.before), u64::from(request.after))
        } else {
            (u64::from(request.context), u64::from(request.context))
        };
        let req = sys::SearchRequest {
            struct_size: u32::try_from(std::mem::size_of::<sys::SearchRequest>())
                .expect("request struct fits u32"),
            flags: flags(request),
            max_count: u64::from(request.max_count),
            before_context: before,
            after_context: after,
            pattern: pattern.as_ptr(),
            pattern_len: pattern.len(),
            timeout_ns: run
                .timeout
                .map_or(0, |d| u64::try_from(d.as_nanos()).unwrap_or(u64::MAX)),
            max_results: run.max_results.unwrap_or(0),
            cancel: run.cancel.map_or(std::ptr::null_mut(), |t| t.inner),
        };
        let mut out: *mut sys::irregex_cursor = std::ptr::null_mut();
        let status = {
            let _guard = self.lock.lock().expect("engine lock poisoned");
            unsafe { sys::irregex_search_cursor(self.inner, &req, &mut out) }
        };
        if status != sys::OK {
            return Err(status_error(status, "search"));
        }
        Ok(Cursor {
            inner: out,
            matched: None,
            done: false,
        })
    }
}

impl Drop for Engine {
    fn drop(&mut self) {
        unsafe { sys::irregex_engine_close(self.inner) };
    }
}

/// The `irregex_search_request.flags` bitset for the representable option subset.
fn flags(r: &SearchRequest) -> u32 {
    let mut f = 0;
    if r.fixed {
        f |= sys::FLAG_FIXED;
    }
    if r.ignore_case {
        f |= sys::FLAG_IGNORE_CASE;
    }
    if r.smart_case {
        f |= sys::FLAG_SMART_CASE;
    }
    if r.word {
        f |= sys::FLAG_WORD;
    }
    if r.invert {
        f |= sys::FLAG_INVERT;
    }
    if r.quiet {
        f |= sys::FLAG_QUIET;
    }
    if r.unicode == Some(false) {
        f |= sys::FLAG_NO_UNICODE;
    }
    if r.max_count > 0 {
        f |= sys::FLAG_MAX_COUNT;
    }
    f
}

/// Fail loud for any option the cursor ABI has no field for — the mirror of the
/// engine's own fail-closed `IRREGEX_INVALID` posture, so the in-process face
/// never silently answers a subtly different query than the caller asked.
fn reject_unrepresentable(r: &SearchRequest) -> Result<()> {
    let mut bad: Vec<&str> = Vec::new();
    let flag = |on: bool, name: &'static str, v: &mut Vec<&'static str>| {
        if on {
            v.push(name);
        }
    };
    flag(r.hidden, "hidden", &mut bad);
    flag(r.no_ignore, "no_ignore", &mut bad);
    flag(r.follow, "follow", &mut bad);
    flag(r.no_index, "no_index", &mut bad);
    flag(r.multiline, "multiline", &mut bad);
    flag(r.multiline_dotall, "multiline_dotall", &mut bad);
    flag(r.max_depth != 0, "max_depth", &mut bad);
    flag(
        !r.globs.is_empty()
            || !r.iglobs.is_empty()
            || !r.types.is_empty()
            || !r.not_types.is_empty(),
        "glob/type scoping",
        &mut bad,
    );
    flag(!r.extra_flags.is_empty(), "extra_flags", &mut bad);
    if r.engine != SearchEngine::Linear {
        bad.push("engine=pcre2/auto");
    }
    if bad.is_empty() {
        return Ok(());
    }
    Err(Error::Unrepresentable(format!(
        "the in-process cursor cannot honor {}; use SearchRequest::run for the full CLI surface",
        bad.join(", ")
    )))
}

/// Copy one borrowed [`sys::MatchView`] into an owned [`Match`].
fn to_match(v: &sys::MatchView) -> Match {
    let mut text = decode(v.line, v.line_len);
    // The engine's line view excludes '\n' but may retain a trailing '\r'; strip
    // it to match the cold `--json` parser exactly.
    if text.ends_with('\r') {
        text.pop();
    }
    let subs = if v.submatches.is_null() {
        Vec::new()
    } else {
        let views = unsafe { std::slice::from_raw_parts(v.submatches, v.nsubmatches) };
        views
            .iter()
            .map(|s| Submatch {
                text: decode(s.text, s.len),
                start: s.start,
                end: s.end,
            })
            .collect()
    };
    Match {
        path: decode(v.path, v.path_len),
        line_number: v.line_number,
        text,
        kind: if v.kind == 1 {
            MatchKind::Context
        } else {
            MatchKind::Match
        },
        submatches: subs,
    }
}

/// Decode aliased engine bytes as UTF-8 (lossily — invalid bytes become U+FFFD,
/// matching the JSON transport's `String` records).
fn decode(ptr: *const u8, len: usize) -> String {
    if ptr.is_null() || len == 0 {
        return String::new();
    }
    String::from_utf8_lossy(unsafe { std::slice::from_raw_parts(ptr, len) }).into_owned()
}

/// A NUL-checked path → `CString` (a path with an interior NUL is unrepresentable).
fn cstring(p: &Path) -> Result<CString> {
    #[cfg(unix)]
    let bytes = {
        use std::os::unix::ffi::OsStrExt;
        p.as_os_str().as_bytes().to_vec()
    };
    #[cfg(not(unix))]
    let bytes = p.to_string_lossy().into_owned().into_bytes();
    CString::new(bytes).map_err(|_| Error::Unrepresentable(format!("root path has interior NUL: {p:?}")))
}

/// Map a negative native status to the crate's typed error.
fn status_error(status: i32, what: &str) -> Error {
    match status {
        sys::STALE => Error::UnsupportedPattern(format!(
            "{what}: pattern is outside the linear-time engine \
             (use SearchRequest with engine=auto/pcre2 for lookaround)"
        )),
        sys::OOM => Error::Failed(format!("{what}: native out of memory")),
        sys::OPEN_FAILED => Error::Failed(format!("{what}: could not stand up the warm corpus")),
        sys::INVALID => Error::Failed(format!("{what}: invalid or wrongly-sized request")),
        other => Error::Failed(format!("{what}: native status {other}")),
    }
}
